core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java

```diff
@@ -162,119 +148,138 @@ public int getTopologyId() {
       return topologyId;
    }
 
-   //todo [anistor] check thread interrupt status in loops to implement faster cancellation
-   public void run() {
+   /**
+    * Starts sending entries from the data container and the first loader with fetch persistent data enabled
+    * to the target node.
+    *
+    * @return a completion stage that completes when all the entries have been sent.
+    */
+   public CompletionStage<Void> execute() {
+      CompletableFuture<Void> taskFuture = new CompletableFuture<>();
       try {
-         // TODO: need to change to SDC.forEachSegment
-         // send data container entries
-         for (InternalCacheEntry ice : dataContainer) {
-            Object key = ice.getKey();  //todo [anistor] should we check for expired entries?
-            int segmentId = keyPartitioner.getSegment(key);
-            if (segments.contains(segmentId) && !ice.isL1Entry()) {
-               InternalCacheEntry entry = mapEntryFromDataContainer.apply(ice, entryFactory);
-               if (entry != null) {
-                  sendEntry(entry, segmentId);
-               }
-            }
+         if (subscription != null) {
+            throw new IllegalStateException("execute() may only be called once once");
          }
-
-         AdvancedCacheLoader<Object, Object> stProvider = persistenceManager.getStateTransferProvider();
-         if (stProvider != null) {
-            try {
-               Flowable.fromPublisher(stProvider.entryPublisher(k -> !dataContainer.containsKey(k), true, true))
-                     .blockingForEach(me -> {
-                        int segmentId = keyPartitioner.getSegment(me.getKey());
-                        if (segments.contains(segmentId)) {
-                           try {
-                              InternalCacheEntry entry = mapEntryFromStore.apply(me, entryFactory);
-                              if (entry != null) {
-                                 sendEntry(entry, segmentId);
-                              }
-                           } catch (CacheException e) {
-                              log.failedLoadingValueFromCacheStore(me.getKey(), e);
-                           }
-                        }
-                     });
-            } catch (CacheException e) {
-               log.failedLoadingKeysFromCacheStore(e);
-            }
+         AtomicReference<List<InternalCacheEntry<Object, Object>>> batchRef =
+            new AtomicReference<>(Collections.emptyList());
+         subscription =
+            Flowable.concat(publishDataContainerEntries(), publishStoreEntries())
+                    .buffer(chunkSize)
+                    .concatMapCompletable(
```

Comment:
I took a peek and this method does a prefetch of 2 elements. I believe this is fine, but I just wanted to point that out.

Reply:
Yeah, I remember reading how it requests new elements when the prefetch queue is half-depleted, so I figured 2 is about the same as 1. But now that you mentioned it I thought more about it and changed it to 1 to limit memory usage.